### DISCUSSION TITLE
Can I use Atom setter in useMutationQuery onSuccess method

### DISCUSSION THREAD

**Discussion by shongs27** [pmndrs/jotai#2338] [Q&A]
```ts
const testAtom = atom(null);

export const [, changeCoach] = atomsWithMutation<
  AxiosResponse,
  AxiosError,
  ICoachWindow.CoachRequest
>((get) => ({
  mutationKey: ['coachPatch', get(companyIdsAtom)],
  mutationFn: async ({ uid, req }) => {
    const companyIds = encodeURIComponent(get(companyIdsAtom).join(','));
    const headers = get(headerInfo);

    return await fetcher({
      ...headers,
      method: 'patch',
      url: `/company/${companyIds}/user/${uid}/coach`,
      data: req,
    });
  },
  onSuccess: () => {
    get(queryClientAtom).invalidateQueries([
      `/company/${get(companyIdsAtom).join(',')}/user/target`,
      get(targetInquireConditionsAtoms),
      get(headerInfo),
    ]);
   set(testAtom, true) // => ???? how to set
  },
}));
```
What I want is Atom setter in onSuccess method
Without useAtom, How can I make change Atom in onSuccess method

or how to get to solve? thanks for reading my question

**Comment by dai-shi** [comment]
I suppose you are using the old version of `jotai-tanstack-query`.
@kalijonn might be able to help with the new version.

**Comment by kalijonn** [comment]
There are 2 typical patterns to handle side effects.
1. Invalidating queries. I think you're already doing this. This is useful when the state you're trying to manage is within tanstack query.
2. [jotai-effect](https://jotai.org/docs/extensions/effect)

If both of these don't work for you, please elaborate on what you are trying to do. What the test atom is supposed to do etc. This would help see if an existing pattern can help with your scenario.

**Reply by mengxi-ream** [reply]
Could you give some examples for using setter in onSuccess? I need something like:
```ts
export const explainAtom = atomWithMutation(() => ({
  mutationKey: ["explainArticle"],
  mutationFn: mutationFn,
  onSuccess: () => {
    set(readStateAtom, "Explaining"); // how to do this?
  },
  onError: () => {
    toast.error("Failed to generate the explanation");
  },
}));

type ReadState =
  | "Idle"
  | "Analyzing"
  | "Continue?"
  | "Explaining"

export const readStateAtom = atom<ReadState>("Idle");
```
